tools/c7n_azure/c7n_azure/resources/app_gateway.py

```diff
@@ -33,3 +36,56 @@ class resource_type(ArmResourceManager.resource_type):
             'resourceGroup'
         )
         resource_type = 'Microsoft.Network/applicationGateways'
+
+
+@ApplicationGateway.filter_registry.register('waf')
+class ApplicationGatewayWafFilter(Filter):
+    """
+    Filter Application Gateways using WAF rule configuration
+
+    :example:
+
+    Return all the App Gateways which have rule '944240' disabled.
+
+    .. code-block:: yaml
+
+        policies:
+          - name: test-app-gateway
+            resource: azure.application-gateway
+            filters:
+              - type: waf
+                override_rule: 944240
+                state: disabled
+    """
+    
+    schema = type_schema(
+        'waf',
+        required=['override_rule', 'state'],
+        **{
+            'override_rule': {'type': 'number'},
+            'state': {'type': 'string', 'enum': ['disabled']}}
+    )
+
+    def process(self, resources, event=None):
+
+        filter_override_rule = self.data.get('override_rule')
+        filter_state = self.data.get('state')
+
+        client = self.manager.get_client()
+        result = []
+
+        for resource in resources:
+            if 'firewallPolicy' in resource['properties']:
+                waf_policy_name = resource['properties']['firewallPolicy']['id'].split('/')[-1]
+
+                app_gate_waf = client.web_application_firewall_policies.\
```

Comment:
ideally we wouldn't fetch the same policy multiple times.

Reply:
Refactored the code to fetch the all the waf policies once.